// wechaty-puppet-service/dist/src/config.js
"use strict";
/// <reference path="./typings.d.ts" />
Object.defineProperty(exports, "__esModule", { value: true });
exports.GET_WECHATY_SERVICE_DISCOVERY_ENDPOINT = exports.GET_WECHATY_PUPPET_SERVICE_TOKEN = exports.GET_WECHATY_PUPPET_SERVICE_ENDPOINT = exports.VERSION = exports.GRPC_OPTIONS = exports.log = void 0;
const wechaty_puppet_1 = require("wechaty-puppet");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return wechaty_puppet_1.log; } });
const version_1 = require("./version");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_1.VERSION; } });
const GRPC_OPTIONS = {
// https://github.com/wechaty/wechaty-puppet-service/issues/86
// 'grpc.max_receive_message_length': 1024 * 1024 * 150,
// 'grpc.max_send_message_length': 1024 * 1024 * 150,
};
exports.GRPC_OPTIONS = GRPC_OPTIONS;
// Huan(202011): use a function to return the value in time.
const GET_WECHATY_PUPPET_SERVICE_TOKEN = (token) => {
    if (token) {
        return token;
    }
    if (process.env['WECHATY_PUPPET_SERVICE_TOKEN']) {
        return process.env['WECHATY_PUPPET_SERVICE_TOKEN'];
    }
    /**
     * Huan(202102): remove this deprecated warning after Dec 31, 2021
     */
    if (process.env['WECHATY_PUPPET_HOSTIE_TOKEN']) {
        wechaty_puppet_1.log.warn('wechaty-puppet-service', [
            '',
            'WECHATY_PUPPET_HOSTIE_TOKEN has been deprecated,',
            'please use WECHATY_PUPPET_SERVICE_TOKEN instead.',
            'See: https://github.com/wechaty/wechaty-puppet-service/issues/118',
            '',
        ].join(' '));
        return process.env['WECHATY_PUPPET_HOSTIE_TOKEN'];
    }
    return undefined;
};
exports.GET_WECHATY_PUPPET_SERVICE_TOKEN = GET_WECHATY_PUPPET_SERVICE_TOKEN;
const GET_WECHATY_PUPPET_SERVICE_ENDPOINT = (endpoint) => {
    if (endpoint) {
        return endpoint;
    }
    if (process.env['WECHATY_PUPPET_SERVICE_ENDPOINT']) {
        return process.env['WECHATY_PUPPET_SERVICE_ENDPOINT'];
    }
    /**
     * Huan(202102): remove this deprecated warning after Dec 31, 2021
     */
    if (process.env['WECHATY_PUPPET_HOSTIE_ENDPOINT']) {
        wechaty_puppet_1.log.warn('wechaty-puppet-service', [
            '',
            'WECHATY_PUPPET_HOSTIE_ENDPOINT has been deprecated,',
            'please use WECHATY_PUPPET_SERVICE_ENDPOINT instead.',
            'See: https://github.com/wechaty/wechaty-puppet-service/issues/118',
            '',
        ].join(' '));
        return process.env['WECHATY_PUPPET_HOSTIE_ENDPOINT'];
    }
    return undefined;
};
exports.GET_WECHATY_PUPPET_SERVICE_ENDPOINT = GET_WECHATY_PUPPET_SERVICE_ENDPOINT;
const GET_WECHATY_SERVICE_DISCOVERY_ENDPOINT = (endpoint) => {
    if (endpoint) {
        return endpoint;
    }
    return process.env['WECHATY_SERVICE_DISCOVERY_ENDPOINT']
        || 'https://api.chatie.io';
};
exports.GET_WECHATY_SERVICE_DISCOVERY_ENDPOINT = GET_WECHATY_SERVICE_DISCOVERY_ENDPOINT;
