// wechaty-puppet/dist/src/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.StateSwitch = exports.MemoryCard = exports.FileBox = exports.envVars = exports.log = void 0;
const brolog_1 = require("brolog");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return brolog_1.log; } });
const file_box_1 = require("file-box");
Object.defineProperty(exports, "FileBox", { enumerable: true, get: function () { return file_box_1.FileBox; } });
const memory_card_1 = require("memory-card");
Object.defineProperty(exports, "MemoryCard", { enumerable: true, get: function () { return memory_card_1.MemoryCard; } });
const state_switch_1 = require("state-switch");
Object.defineProperty(exports, "StateSwitch", { enumerable: true, get: function () { return state_switch_1.StateSwitch; } });
const version_1 = require("./version");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_1.VERSION; } });
const envVars = __importStar(require("./env-vars"));
exports.envVars = envVars;
const logLevel = process.env['WECHATY_LOG'];
if (logLevel) {
    brolog_1.log.level(logLevel.toLowerCase());
    brolog_1.log.silly('Puppet', 'Config: WECHATY_LOG set level to %s', logLevel);
}
